### ISSUE TITLE
Super doesn't work inside applications

### ISSUE THREAD

**Issue by bartvm** [bartvm/blocks#475]
I think this is because we're replacing the methods with `BoundApplication`. Instead we should probably override `__get__` to return a `BoundApplication`.

**Comment by rizar** [comment]
Sure? The code below works just fine. 

```
from blocks.bricks.base import Brick, application

class A(Brick):

    def apply(self):
        print "A"

class B(A):

    @application
    def apply(self):
        print "B"

b = B()
b.apply()
super(B, b).apply()
```

**Comment by bartvm** [comment]
Don't have time to do more digging, but this is the example that I was hoping would work, but doesn't:

```python
from blocks.bricks.recurrent import GatedRecurrent
from blocks.bricks.parallel import Fork
from blocks.bricks import Tanh

class GatedRecurrentWithContext(GatedRecurrent):
    @application(states=['states'], outputs=['states'],
                 contexts=['context'])
    def apply(self, context, *args, **kwargs):
        kwargs['inputs'] += context
        return super(GatedRecurrentWithContext, self).apply(*args, **kwargs)
        
        
fork = Fork(input_dim=100, output_dims=[100] * 3,
            output_names=['inputs', 'reset_inputs', 'update_inputs'])

gated_recurrent = GatedRecurrentWithContext(Tanh(), None, 100)

inputs = tensor.tensor3('inputs')
context = tensor.matrix('context').dimshuffle('x', 0, 1)
gated_recurrent.apply(context, **fork.apply(inputs, as_dict=True))
```

**Comment by rizar** [comment]
That was because of my hack in `recurrent`, sorry. See #512, if you like it I will add a regression test and merge.

**Comment by dwf** [comment]
This was closed via #512.
